File: Lessons/Lesson 1 - Iterations/BinaryGap/java/BinaryGap.java
// you can also use imports, for example:
// import java.util.*;

// you can write to stdout for debugging purposes, e.g.
// System.out.println("this is a debug message");

import java.util.LinkedList;
import java.util.List;

class Solution {
    public int solution(int N) {
        List<Integer> binNum = new LinkedList<>();
        while (N > 0) {
            if (N % 2 == 1) {
                binNum.add(1);
            }
            else {
                binNum.add(0);
            }
            N /= 2;
        }
        boolean flag = false;
        int result = 0;
        int temp = 0;
        for (int i = 0; i < binNum.size(); i++) {
            if (flag == false && binNum.get(i) == 1) {
                flag = true;
            }
            else if (flag == true && binNum.get(i) == 1) {
                if (temp > result) {
                    result = temp;
                }
                temp = 0;                
            }
            else if (flag == true && binNum.get(i) == 0) {
                temp++;
            }
        }
        return result;
    }
}

File: Lessons/Lesson 1 - Iterations/BinaryGap/c++/BinaryGap.cpp
int solution(int N) {
    
    string binNum = "";
    while (N > 0) {
        if (N % 2 == 0) {
            binNum = '0' + binNum;
        }
        else {
            binNum = '1' + binNum;
        }
        N /= 2;
    }

    int res = 0;
    int temp = 0;
    bool flag = false;

    for (int i = 0; i < binNum.length(); i++) {
        
        if (binNum[i] == '1' && flag == false) {
            flag = true;
        }
        else if (binNum[i] == '0' && flag == false) {
            continue;
        }
        else if (binNum[i] == '0' && flag == true) {
            temp++;
        }
        else {          // binNum[i] == '1' && flag == true
            if (temp > res) {
                res = temp;
            }
            temp = 0;
        }
    }
    return res;
}